File: frontend/src/api/push.ts
import type { PushCreate, PushInfo } from "sdk";
import type { Api } from "../api.tsx";

export class Push {
	public api: Api = null as unknown as Api;

	async register(body: PushCreate): Promise<PushInfo> {
		const { data, error } = await this.api.client.http.POST(
			"/api/v1/push",
			{
				body,
			},
		);
		if (error) {
			console.error(error);
			throw new Error(error);
		}
		return data as PushInfo;
	}

	async delete(): Promise<void> {
		const { error } = await this.api.client.http.DELETE(
			"/api/v1/push",
		);
		if (error) {
			console.error(error);
			throw new Error(error);
		}
	}

	async get(): Promise<PushInfo> {
		const { data, error } = await this.api.client.http.GET(
			"/api/v1/push",
		);
		if (error) {
			console.error(error);
			throw new Error(error);
		}
		return data as PushInfo;
	}
}

File: crate-common/src/v1/routes/push.rs
use lamprey_macros::endpoint;

/// Push register
///
/// Register web push for this session
#[endpoint(
    post,
    path = "/push",
    tags = ["push"],
    scopes = [Full],
    response(OK, body = PushInfo, description = "ok"),
)]
pub mod push_register {
    use crate::v1::types::push::{PushCreate, PushInfo};

    pub struct Request {
        #[json]
        pub push: PushCreate,
    }

    pub struct Response {
        #[json]
        pub push: PushInfo,
    }
}

/// Push delete
///
/// Remove web push for this session
#[endpoint(
    delete,
    path = "/push",
    tags = ["push"],
    scopes = [Full],
    response(NO_CONTENT, description = "ok"),
)]
pub mod push_delete {
    pub struct Request {}
    pub struct Response {}
}

/// Push get
///
/// Get web push subscription for this session
#[endpoint(
    get,
    path = "/push",
    tags = ["push"],
    scopes = [Full],
    response(OK, body = PushInfo, description = "ok"),
)]
pub mod push_get {
    use crate::v1::types::push::PushInfo;

    pub struct Request {}

    pub struct Response {
        #[json]
        pub push: PushInfo,
    }
}
